File: Chapter 2/Section 2.3/zerosum/zerosum.py
"""
ID: ten.to.1
TASK: zerosum
LANG: PYTHON3
"""

#!/usr/bin/python

file_in  = open("zerosum.in", "r")
file_out = open("zerosum.out", "w")

N = int(file_in.read())

choose = [" ", "+", "-"]

def solve(b, n):
	if(n == N):
		b += str(N)
		if(eval(b.replace(" ", "")) == 0):
			file_out.write(b + "\n")
		return
	
	for c in choose:
		solve(b + str(n) + c, n + 1)

solve("", 1)

File: Chapter 2/Section 2.3/zerosum/zerosum.cpp
/*
ID: ten.to.1
TASK: zerozum
LANG: C++
*/

#include <iostream>
#include <fstream>

void solve(std::string b, int n, int sum, const int& N)
{
	char choose[] = {'+', '-', ' '};
	
	if(n == N)
	{
		b += std::to_string(N);
		
		std::cout << b << '\n';
		
		return;
	}
	
	for(int i = 0; i < 3; i++)
	{
		solve(b + std::to_string(n) + choose[i], n + 1, sum, N);
	}
}

int main()
{
	std::ifstream file_in("zerosum.in");
	std::ofstream file_out("zersum.out");
	
	int N; file_in >> N;
	
	solve("", 1, 0, N);
}

